spec/concepts/class_member/create_spec.rb

```
@@ -5,61 +5,138 @@
-      response = described_class.call(school_class:, class_member_params:)
-      expect(response.failure?).to be(true)
-    end
+      it 'does not create a class member' do
+        expect { described_class.call(school_class:, students:) }.not_to change(ClassMember, :count)
+      end
 
-    it 'returns the error message in the operation response' do
-      response = described_class.call(school_class:, class_member_params:)
-      expect(response[:error]).to match(/Error creating class member/)
+      it 'returns a failed operation response' do
+        response = described_class.call(school_class:, students:)
+        expect(response.failure?).to be(true)
+      end
+
+      it 'returns the error message in the operation response' do
+        response = described_class.call(school_class:, students:)
+        expect(response[:error]).to match(/No valid students provided/)
+      end
+
+      it 'sent the exception to Sentry' do
+        described_class.call(school_class:, students:)
+        expect(Sentry).to have_received(:capture_exception).with(kind_of(StandardError))
+      end
     end
 
-    it 'sent the exception to Sentry' do
-      described_class.call(school_class:, class_member_params:)
-      expect(Sentry).to have_received(:capture_exception).with(kind_of(StandardError))
+    context 'with a student from a different school' do
+      let(:different_school) { create(:school) }
+      let(:different_school_student) { create(:student, school: different_school) }
+
+      context 'with non existent students' do
+        let(:students) { [different_school_student] }
+
+        it 'does not create a class member' do
+          expect { described_class.call(school_class:, students:) }.not_to change(ClassMember, :count)
+        end
+
+        it 'returns a successful operation response' do
+          response = described_class.call(school_class:, students:)
+          expect(response.success?).to be(true)
+        end
```

Why a successful response here?

---

Is it so we can surface the errors from each student?

---

yes its exactly that. So rails will pick up if there's a `response.error` however not `response.errors` so I'm adding to the former for complete request failures but given its a batch endpoint I'm adding to `errors` for the individual cases that error. A little implicit but I think it makes sense and in combination with an empty `class_members` array and `errors` being returned we should be able to surface and expand on this on the frontend.

Let me know what you think though 👍

---

I think that makes sense, let's go with that, and we can always change it if it surfaces some weird edge cases 👍